src/components/Viewport.ts

```diff
@@ -709,12 +724,58 @@ export default class Viewport {
   public calcVisiblePanels(): Panel[] {
     const allPanels = this.panelManager.allPanels();
     if (this.options.renderOnlyVisible) {
-      const { min, max } = this.state.visibleIndex;
-      const visiblePanels = min >= 0
-        ? allPanels.slice(min, max + 1)
-        : allPanels.slice(0, max + 1).concat(allPanels.slice(min));
+      const cameraPos = this.getCameraPosition();
+      const viewportSize = this.getSize();
+      const basePanel = this.nearestPanel!;
 
-      return visiblePanels.filter(panel => panel);
+      const getNextPanel = (panel: Panel) => {
+        const nextPanel = panel.nextSibling;
+
+        if (nextPanel && nextPanel.getPosition() >= panel.getPosition()) {
+          return nextPanel;
+        } else {
+          return null;
+        }
+      };
+
+      const getPrevPanel = (panel: Panel) => {
+        const prevPanel = panel.prevSibling;
+
+        if (prevPanel && prevPanel.getPosition() <= panel.getPosition()) {
+          return prevPanel;
+        } else {
+          return null;
+        }
+      };
+
+      const isOutOfBoundNext = (panel: Panel) => panel.getPosition() >= cameraPos + viewportSize;
+      const isOutOfBoundPrev = (panel: Panel) => panel.getPosition() + panel.getSize() <= cameraPos;
+
+      const getVisiblePanels = (
+        panel: Panel,
+        getNext: (panel: Panel) => Panel | null,
+        isOutOfViewport: (panel: Panel) => boolean,
+      ): Panel[] => {
+        const visiblePanels: Panel[] = [];
+
+        let lastPanel = panel;
+        while (true) {
+          const nextPanel = getNext(lastPanel);
+          if (!nextPanel || isOutOfViewport(nextPanel)) {
+            break;
+          }
+          visiblePanels.push(nextPanel);
+          lastPanel = nextPanel;
+        }
+        return visiblePanels;
+      };
+
+      const panelCount = this.panelManager.getPanelCount();
+      const getAbsIndex = (panel: Panel) => panel.getIndex() + (panel.getCloneIndex() + 1) * panelCount;
+      const nextPanels = getVisiblePanels(basePanel, getNextPanel, isOutOfBoundNext);
+      const prevPanels = getVisiblePanels(basePanel, getPrevPanel, isOutOfBoundPrev);
+
+      return [basePanel, ...nextPanels, ...prevPanels].sort((panel1, panel2) => getAbsIndex(panel1) - getAbsIndex(panel2));
```

Comment:
@WoodNeck 

I've been thinking about this code and it looks like the code below is not a condition?
```js
return allPanels.filter(panel => {
        const outsetProgress = panel.getOutsetProgress();
        return outsetProgress > -1 && outsetProgress < 1;
      });
```

Reply:
You're right. That should work for the non-renderOnlyVisible cases.
I'll see that works.

Reply:
I figured why. That's basically more efficient to do that.
Filtering all panels can be unnecessary when cases like there're plenty of panels but only 3 panels are visible.